File: Novomber 2022/Day 2/C++ Code/Enemy.cpp
// Problem Link :- https://practice.geeksforgeeks.org/problems/enemy/1
// Video Solution Link :- https://www.youtube.com/watch?v=oer2YIPWDgI


//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++

class Solution
{
    // Time Complexity :- O(k.log(k));
    // Space Complexity :- O(k);
    public:
        int largestArea(int n,int m,int k,vector<vector<int>> &enemy)
        {
            //  code here
            if(k==0) return m*n;
            vector<int> col;
            vector<int> row;
            for(vector<int> a : enemy) {
                row.push_back(a[0]);
                col.push_back(a[1]);
            }
            sort(col.begin(),col.end());
            int max_column=col[0]-1;
            for(int i=0;i<col.size();i++) {
                if(i==col.size()-1) {
                    max_column = max(max_column,(m+1)-col[i]-1);
                } else {
                    max_column = max(max_column,col[i+1]-col[i]-1);
                }
            }
            sort(row.begin(),row.end());
            int max_row = row[0]-1;
            for(int i=0;i<row.size();i++) {
                if(i==enemy.size()-1) {
                    max_row = max(max_row,(n+1)-row[i]-1);
                } else {
                    max_row = max(max_row,row[i+1]-row[i]-1);
                }
            }
            return max_row*max_column;
        }
};

//{ Driver Code Starts.

int main()
{
    int t;
    cin>>t;
    while(t--)
    {
        int n,m;
        cin>>n>>m;
        int k;
        cin>>k;
        vector<vector<int>> e(k,vector<int>(2));
        for(int i=0;i<k;i++)
            cin>>e[i][0]>>e[i][1];
        Solution a;
        cout<<a.largestArea(n,m,k,e)<<endl;
    }
    return 0;
}
// } Driver Code Ends

File: June 2022/Day 29/Enemy/enemy.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;


class IntArray
{
    public static int[] input(BufferedReader br, int n) throws IOException
    {
        String[] s = br.readLine().trim().split(" ");
        int[] a = new int[n];
        for(int i = 0; i < n; i++)
            a[i] = Integer.parseInt(s[i]);
        
        return a;
    }
    
    public static void print(int[] a)
    {
        for(int e : a)
            System.out.print(e + " ");
        System.out.println();
    }
    
    public static void print(ArrayList<Integer> a)
    {
        for(int e : a)
            System.out.print(e + " ");
        System.out.println();
    }
}


class IntMatrix
{
    public static int[][] input(BufferedReader br, int n, int m) throws IOException
    {
        int[][] mat = new int[n][];
        
        for(int i = 0; i < n; i++)
        {
            String[] s = br.readLine().trim().split(" ");
            mat[i] = new int[s.length];
            for(int j = 0; j < s.length; j++)
                mat[i][j] = Integer.parseInt(s[j]);
        }
        
        return mat;
    }
    
    public static void print(int[][] m)
    {
        for(var a : m)
        {
            for(int e : a)
                System.out.print(e + " ");
            System.out.println();
        }
    }
    
    public static void print(ArrayList<ArrayList<Integer>> m)
    {
        for(var a : m)
        {
            for(int e : a)
                System.out.print(e + " ");
            System.out.println();
        }
    }
}

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while(t-- > 0){
            
            int[] a = IntArray.input(br, 3);
            int n = a[0],m=a[1],k=a[2];
            
            int[][] e = IntMatrix.input(br, a[2], 2);
            
            Solution obj = new Solution();
            int res = obj.largestArea(n,m,k,e);
            
            System.out.println(res);
            
        }
    }
}

// } Driver Code Ends




       

class Solution {
    
    // Time Complexity :- O(k.log(k));
    // Space Complexity :- O(k);
    public static int largestArea(int n,int m,int k,int[][] enemy) {
        // code here
        if(k==0) return m*n;
        
        Arrays.sort(enemy,(a,b) ->(a[1]-b[1]));
        int max_column=enemy[0][1]-1;
        for(int i=0;i<enemy.length;i++) {
            if(i==enemy.length-1) {
                max_column = Math.max(max_column,(m+1)-enemy[i][1]-1);
            } else {
                max_column = Math.max(max_column,enemy[i+1][1]-enemy[i][1]-1);
            }
        }
        
        Arrays.sort(enemy,(a,b) -> (a[0]-b[0]));
        int max_row = enemy[0][0]-1;
        for(int i=0;i<enemy.length;i++) {
            if(i==enemy.length-1) {
                max_row = Math.max(max_row,(n+1)-enemy[i][0]-1);
            } else {
                max_row = Math.max(max_row,enemy[i+1][0]-enemy[i][0]-1);
            }
        }
        return max_row*max_column;
    }
}
 
